### PR TITLE
Adding capability for Echo Show #159

### PR DESCRIPTION
**Pull request by johnddias**
in models.py there are now:
     display_render - for BodyTemplate
     list_display_render - for ListTemplate

 I have tested this with BodyTemplate1, BodyTemplate2 and ListTemplate1.  Other templates will likely require modification.

  in core.py there is now:
     display_element_selected - a decorator for touch selection requests from Echo Show.  Tokens are your friend here.

### CODE REVIEW

File: flask_ask/core.py [hunk 1]
```diff
@@ -75,20 +67,15 @@ def __init__(self, app=None, route=None, blueprint=None, stream_cache=None, path
...
+            self.init_app(app)
         elif blueprint is not None:
-            self.init_blueprint(blueprint, path)
-        if stream_cache is None:
-            self.stream_cache = SimpleCache()
-        else:
-            self.stream_cache = stream_cache
+            self.init_blueprint(blueprint)
 
-    def init_app(self, app, path='templates.yaml'):
```
Comment (johnwheeler) [inline flask_ask/core.py]: This removes the ability for the client to specify a different template. Can you tell me your thoughts on this change @johndias?

File: flask_ask/models.py [hunk 2]
```diff
@@ -209,11 +268,11 @@ def _audio_item(self, stream_url=None, offset=0, push_buffer=True):
         # new stream
         else:
             stream['url'] = stream_url
-            stream['token'] = str(uuid.uuid4())
+            stream['token'] = str(random.randint(10000, 100000))
```
Comment (johnwheeler) [inline flask_ask/models.py]: if the token must be unique, this no longer ensures it. i don't know much about the stream handling code, but this would appear to be a breaking change if uniqueness was required originally. thoughts?

File: flask_ask/core.py [hunk 3]
```diff
@@ -198,7 +183,7 @@ def session_ended(self, f):
 
         @ask.session_ended
         def session_ended():
-            return "{}", 200
+            return question(), 200
```
Comment (johnwheeler) [inline flask_ask/core.py]: to me, it makes more sense to return empty json here rather than question. a lot of people have requested returning question here, but i've always thought empty json was sufficient. what does returning question enhance?

File: flask_ask/core.py [hunk 4]
```diff
@@ -249,9 +234,20 @@ def wrapper(*args, **kw):
             return f
         return decorator
 
-    def default_intent(self, f):
-        """Decorator routes any Alexa IntentRequest that is not matched by any existing @ask.intent routing."""
```
Comment (johnwheeler) [inline flask_ask/core.py]: why was this nixed?

File: flask_ask/models.py [hunk 5]
```diff
@@ -84,13 +83,73 @@ def link_account_card(self):
         card = {'type': 'LinkAccount'}
         self._response['card'] = card
         return self
-        
-    def consent_card(self, permissions):
```
Comment (johnwheeler) [inline flask_ask/models.py]: why was this removed?

**Comment by johnddias** [comment]
Oh my, I don't recognize any of those changes.  I made a fresh fork, cloned
it and added my code.

Let me recheck my fork, something is wrong.

On Jul 22, 2017 9:00 AM, "John Wheeler" <notifications@github.com> wrote:

*@johnwheeler* requested changes on this pull request.
------------------------------

In flask_ask/core.py
<https://github.com/johnwheeler/flask-ask/pull/163#discussion_r128897255>:

>
-    def init_app(self, app, path='templates.yaml'):

This removes the ability for the client to specify a different template.
Can you tell me your thoughts on this change @johndias?
------------------------------

In flask_ask/models.py
<https://github.com/johnwheeler/flask-ask/pull/163#discussion_r128897294>:

> @@ -209,11 +268,11 @@ def _audio_item(self, stream_url=None, offset=0, push_buffer=True):
         # new stream
         else:
             stream['url'] = stream_url
-            stream['token'] = str(uuid.uuid4())
+            stream['token'] = str(random.randint(10000, 100000))

if the token must be unique, this no longer ensures it. i don't know much
about the stream handling code, but this would appear to be a breaking
change if uniqueness was required originally. thoughts?
------------------------------

In flask_ask/core.py
<https://github.com/johnwheeler/flask-ask/pull/163#discussion_r128897326>:

> @@ -198,7 +183,7 @@ def session_ended(self, f):

         @ask.session_ended
         def session_ended():
-            return "{}", 200
+            return question(), 200

to me, it makes more sense to return empty json here rather than question.
a lot of people have requested returning question here, but i've always
thought empty json was sufficient. what does returning question enhance?
------------------------------

In flask_ask/core.py
<https://github.com/johnwheeler/flask-ask/pull/163#discussion_r128897347>:

> @@ -249,9 +234,20 @@ def wrapper(*args, **kw):
             return f
         return decorator

-    def default_intent(self, f):
-        """Decorator routes any Alexa IntentRequest that is not
matched by any existing @ask.intent routing."""

why was this nixed?
------------------------------

In flask_ask/models.py
<https://github.com/johnwheeler/flask-ask/pull/163#discussion_r128897367>:

> @@ -84,13 +83,73 @@ def link_account_card(self):
         card = {'type': 'LinkAccount'}
         self._response['card'] = card
         return self
-
-    def consent_card(self, permissions):

why was this removed?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/johnwheeler/flask-ask/pull/163#pullrequestreview-51623190>,
or mute the thread

**Comment by johnwheeler** [comment]
Thank you!

**Comment by sodle** [comment]
@johnddias Can you provide an example of how to handle a display_element_selected event? Documentation seems a bit lacking.

**Comment by johnddias** [comment]
@sodle Basically it just detects that there has been an element selected from a display list (i.e. ListTemplate1).  You'll want to use tokens in the list items.  For example:

```def interactive_resp(data):
    if session.attributes["CurrentTree"] == "GroupedAlerts":
        listItems = []
        resDict = session.attributes["resourceDict"]
        for res in resDict:
            listItem = {
                "token":resDict[res],
                "textContent": {
                    "primaryText": {
                        "text":res,
                        "type":"PlainText"
                    }
                }
            }
            listItems.append(listItem)


    enhancedResponse = question("Here are the impacted objects.").list_display_render(template="ListTemplate1", title="Impacted Objects", backButton="VISIBILE", token=None, \
    background_image_url=render_template('backgroundImageURL'), listItems=listItems)

    return enhancedResponse
```
You can see that I'm setting a token for each list item.  You will also want to make sure that you check that you're responding to a Echo Show, using the context.System.device.supportedInterfaces.Display attribute in the request.

```
 if ((session.attributes["CurrentTree"] == "GroupedAlerts") and (context.System.device.supportedInterfaces.Display)):
        enhancedResponse = interactive_resp(outputSpeech)
        return enhancedResponse
```
Once your list is displayed, if the user selects an item from the list you can use display_element_selected to capture that request (and again, this is where tokens come in very handy).

```
@ask.display_element_selected

def element():
    fullResponse = on_element_select(request["token"])
    return fullResponse
```

**Comment by sodle** [comment]
@johnddias Ah, ok, that's what I thought. But I've discovered a bug where the `_display_element_selected_func` never gets called when a request is received. I've added a pull request to fix.
